### PR TITLE
[LoongArch] Refactor LoongArchABI::computeTargetABI

### PR DESCRIPTION
**Pull request by wangleiat**
The previous logic did not consider whether the architectural features
meet the requirements of the ABI, resulting in the generation of
incorrect object files in some cases. For example:

```
llc -mtriple=loongarch64 -filetype=obj test/CodeGen/LoongArch/ir-instruction/fadd.ll -o t.o
llvm-readelf -h t.o
```
The object file indicates the ABI as lp64d, however, the generated code
is lp64s.

The new logic introduces the `feature-implied` ABI. When both target-abi
and triple-implied ABI are invalid, the feature-implied ABI is used.

### CODE REVIEW

**Comment by wangleiat** [comment]
@xen0n

File: llvm/lib/Target/LoongArch/MCTargetDesc/LoongArchBaseInfo.cpp [hunk 1]
```diff
@@ -52,63 +53,129 @@ static ABI checkABIStandardized(ABI Abi) {
...
+    if (TT.hasEnvironment() && ArgProvidedABI != TripleABI)
+      errs()
+          << "warning: triple-implied ABI conflicts with provided target-abi '"
+          << ABIName << "', using target-abi\n";
+    return checkABIStandardized(ArgProvidedABI);
+  }
+
+  // 2. If the triple-implied ABI is valid, use it.
+  if (IsValidABI(TripleABI)) {
+    // If not specifie target-abi, use the valid triple-implied ABI.
```
Comment (xen0n) [inline llvm/lib/Target/LoongArch/MCTargetDesc/LoongArchBaseInfo.cpp:118]:
```suggestion
    // If target-abi is not specified, use the valid triple-implied ABI.
```

File: llvm/lib/Target/LoongArch/MCTargetDesc/LoongArchBaseInfo.cpp [hunk 2]
```diff
@@ -52,63 +53,129 @@ static ABI checkABIStandardized(ABI Abi) {
...
+      return checkABIStandardized(TripleABI);
 
-  switch (ArgProvidedABI) {
-  case LoongArchABI::ABI_Unknown:
-    // Fallback to the triple-implied ABI if ABI name is not specified or
-    // invalid.
-    if (!ABIName.empty())
+    switch (ArgProvidedABI) {
+    case ABI_Unknown:
+      // Fallback to the triple-implied ABI if ABI name is specified and
```
Comment (xen0n) [inline llvm/lib/Target/LoongArch/MCTargetDesc/LoongArchBaseInfo.cpp:124]:
```suggestion
      // Fallback to the triple-implied ABI if ABI name is specified but
```

File: llvm/lib/Target/LoongArch/MCTargetDesc/LoongArchBaseInfo.cpp [hunk 3]
```diff
@@ -52,63 +53,129 @@ static ABI checkABIStandardized(ABI Abi) {
...
-  case LoongArchABI::ABI_LP64F:
-  case LoongArchABI::ABI_LP64D:
-    if (!Is64Bit) {
-      errs() << "64-bit ABIs are not supported for 32-bit targets, ignoring "
-                "target-abi and using triple-implied ABI\n";
-      return checkABIStandardized(TripleABI);
+    switch (ArgProvidedABI) {
+    case ABI_ILP32F:
+    case ABI_LP64F:
+      errs() << "'" << ABIName
```
Comment (xen0n) [inline llvm/lib/Target/LoongArch/MCTargetDesc/LoongArchBaseInfo.cpp:153]:
```suggestion
      errs() << "warning: the '" << ABIName
```

File: llvm/lib/Target/LoongArch/MCTargetDesc/LoongArchBaseInfo.cpp [hunk 4]
```diff
@@ -52,63 +53,129 @@ static ABI checkABIStandardized(ABI Abi) {
...
+             << "' ABI can't be used for a target that doesn't support the 'F' "
+                "instruction set, ignoring target-abi and using triple-implied "
+                "ABI\n";
+      break;
+    case ABI_ILP32D:
+    case ABI_LP64D:
+      errs() << "'" << ABIName
```
Comment (xen0n) [inline llvm/lib/Target/LoongArch/MCTargetDesc/LoongArchBaseInfo.cpp:160]: same here

File: llvm/lib/Target/LoongArch/MCTargetDesc/LoongArchBaseInfo.cpp [hunk 5]
```diff
@@ -52,63 +53,129 @@ static ABI checkABIStandardized(ABI Abi) {
...
   }
 
-  if (!ABIName.empty() && TT.hasEnvironment() && ArgProvidedABI != TripleABI)
-    errs() << "warning: triple-implied ABI conflicts with provided target-abi '"
-           << ABIName << "', using target-abi\n";
-
-  return checkABIStandardized(ArgProvidedABI);
+  // 3. Parse the 'feature-abi', and use it.
+  if (ABIName.empty())
+    errs() << "The triple-implied ABI is invalid, ignoring triple-implied ABI "
```
Comment (xen0n) [inline llvm/lib/Target/LoongArch/MCTargetDesc/LoongArchBaseInfo.cpp:173]:
```suggestion
    errs() << "warning: the triple-implied ABI is invalid, ignoring "
```

File: llvm/lib/Target/LoongArch/MCTargetDesc/LoongArchBaseInfo.cpp [hunk 6]
```diff
@@ -52,63 +53,129 @@ static ABI checkABIStandardized(ABI Abi) {
...
+              "and using feature-implied ABI\n";
+  else
+    errs() << "The target-abi and triple-implied ABI are invalid, ignoring "
```
Comment (xen0n) [inline llvm/lib/Target/LoongArch/MCTargetDesc/LoongArchBaseInfo.cpp:176]:
```suggestion
    errs() << "warning: both target-abi and the triple-implied ABI are invalid, ignoring "
```

File: llvm/lib/Target/LoongArch/MCTargetDesc/LoongArchBaseInfo.cpp [hunk 7]
```diff
@@ -52,63 +53,129 @@ static ABI checkABIStandardized(ABI Abi) {
...
+              "them and using feature-implied ABI\n";
```
Comment (xen0n) [inline llvm/lib/Target/LoongArch/MCTargetDesc/LoongArchBaseInfo.cpp:177]:
```suggestion
              "and using feature-implied ABI\n";
```

File: llvm/lib/Target/LoongArch/MCTargetDesc/LoongArchBaseInfo.cpp [hunk 8]
```diff
@@ -52,63 +53,129 @@ static ABI checkABIStandardized(ABI Abi) {
...
+  ABI TripleABI = getTripleABI(TT);
+
+  auto GetFeatureABI = [=]() {
+    if (FeatureBits[LoongArch::FeatureBasicD])
+      return Is64Bit ? ABI_LP64D : ABI_ILP32D;
+    if (FeatureBits[LoongArch::FeatureBasicF])
+      return Is64Bit ? ABI_LP64F : ABI_ILP32F;
+    return Is64Bit ? ABI_LP64S : ABI_ILP32S;
+  };
+  auto IsValidABI = [=](ABI Abi) {
```
Comment (xen0n) [inline llvm/lib/Target/LoongArch/MCTargetDesc/LoongArchBaseInfo.cpp:88]: name this `IsABIValidForFeature` for extra clarity?

**Review by xen0n — COMMENTED** [review]
Maybe unify all warning messages' format into `warning: blah blah`? LGTM apart from the natural language usages.

File: llvm/lib/Target/LoongArch/MCTargetDesc/LoongArchBaseInfo.cpp [hunk 9]
```diff
@@ -49,63 +50,127 @@ static ABI checkABIStandardized(ABI Abi) {
...
-             << "' is not a recognized ABI for this target, ignoring and using "
-                "triple-implied ABI\n";
-    return checkABIStandardized(TripleABI);
+ABI computeTargetABI(const Triple &TT, const FeatureBitset &FeatureBits,
+                     StringRef ABIName) {
+  bool Is64Bit = TT.isArch64Bit();
+  ABI ArgProvidedABI = getTargetABI(ABIName);
+  ABI TripleABI = getTripleABI(TT);
+
+  auto GetFeatureABI = [=]() {
```
Comment (SixWeining) [inline llvm/lib/Target/LoongArch/MCTargetDesc/LoongArchBaseInfo.cpp:78]: Maybe move it to where it is called (i.e. line 173) for better readability?

**Review by SixWeining — APPROVED** [review]

**Review by xen0n — APPROVED** [review]

**Pull request merged**